### PR TITLE
bugfix: Add args and options to lazy url block

### PR DESCRIPTION
**Pull request by marblenix**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?
- [x] Have you successfully run `brew man` locally and committed any changes?

-----

See discussion: https://github.com/Homebrew/discussions/discussions/718

tl;dr: `url do...end` blocks currently can't handle options such as `:referrer` or `:cookies`

### CODE REVIEW

File: Library/Homebrew/cask/dsl.rb [hunk 1]
```diff
@@ -181,7 +181,7 @@ def url(*args, **options)
 
       set_unique_stanza(:url, args.empty? && options.empty? && !block_given?) do
         if block_given?
-          LazyObject.new { URL.new(*yield, from_block: true, caller_location: caller_location) }
+          LazyObject.new { URL.new(*yield, *args, **options, from_block: true, caller_location: caller_location) }
```
Comment (reitermarkus) [inline Library/Homebrew/cask/dsl.rb:184]: Does this work?

```suggestion
          url, **options = yield
          LazyObject.new { URL.new(url, **options, from_block: true, caller_location: caller_location) }
```

Comment (marblenix) [reply Library/Homebrew/cask/dsl.rb:184]: I don't think so. I'm not that familiar with `yield` so I'm not sure why this didn't work in the first place. 

```
Error: /usr/local/Homebrew/Library/Homebrew/cask/dsl.rb:184: syntax error, unexpected **arg, expecting '='
          url, **options = yield
               ^~
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
```

Comment (marblenix) [reply Library/Homebrew/cask/dsl.rb:184]: update: this does work!

```ruby
          yielded = *yield
          url = yielded[0]
          opts = yielded[1] || {}
          opts[:from_block] = true
          opts[:caller_location] = caller_location
          LazyObject.new { URL.new(url, opts) }
```

Comment (marblenix) [reply Library/Homebrew/cask/dsl.rb:184]: @reitermarkus would you prefer this s tyle?

Comment (marblenix) [reply Library/Homebrew/cask/dsl.rb:184]: if I do it this way, this pull request can be closed without merging: https://github.com/Homebrew/homebrew-cask/pull/99447

Comment (reitermarkus) [reply Library/Homebrew/cask/dsl.rb:184]: This should work:

```ruby
*args = yield
options = args.last.is_a?(Hash) : args.pop : {}
LazyObject.new { URL.new(*args, **options, from_block: true, caller_location: caller_location) }
```

Comment (marblenix) [reply Library/Homebrew/cask/dsl.rb:184]: fixed, thanks!

**Comment by marblenix** [comment]
Question for the maintainers: Would you prefer i squashed and force pushed to make this a single commit?

**Comment by vitorgalvao** [comment]
@reitermarkus Any blockers?

File: Library/Homebrew/cask/dsl.rb [hunk 2]
```diff
@@ -181,7 +181,11 @@ def url(*args, **options)
 
       set_unique_stanza(:url, args.empty? && options.empty? && !block_given?) do
         if block_given?
-          LazyObject.new { URL.new(*yield, from_block: true, caller_location: caller_location) }
+          LazyObject.new do
+            args = yield
```
Comment (reitermarkus) [inline Library/Homebrew/cask/dsl.rb:185]:
```suggestion
            *args = yield
```

Comment (reitermarkus) [reply Library/Homebrew/cask/dsl.rb:185]: Otherwise `#last` will be called on e.g. a string, as seen in the failing audit currently.

**Comment by reitermarkus** [comment]
> Would you prefer i squashed and force pushed to make this a single commit?

@marblenix, yeah, that would be good. Doesn't necessarily need to be a single commit, but some commit messages could be improved.

File: Library/Homebrew/cask/dsl.rb (hunk 2 above)
Comment (marblenix) [reply Library/Homebrew/cask/dsl.rb]: Thanks! Fixed

**Review by reitermarkus — APPROVED** [review]

**Comment by reitermarkus** [comment]
Thanks, @marblenix, nice work!
